src/bufferwords.jl

```diff
@@ -0,0 +1,180 @@
+mutable struct BufferWord{T} <: AbstractWord{T}
+    storage::Vector{T}
+    lidx::Int
+    ridx::Int
+
+    function BufferWord{T}(v::AbstractVector{<:Integer},
+        freeatbeg=8, freeatend=8) where T
+
+        storage = Vector{T}(undef, freeatbeg + length(v) + freeatend)
+        bw = new{T}(storage, freeatbeg+1, freeatbeg + length(v))
+
+        # placing content in the middle
+        # storage[freeatbeg+1:end-freeatend] .= v
+        @inbounds for i in 1:length(v)
+            bw[i] = v[i]
+        end
+
+        return bw
+    end
+
+    function BufferWord{T}(freeatbeg=8, freeatend=8) where T
+        l = freeatbeg + freeatend
+        return new{T}(Vector{T}(undef, l), freeatbeg+1, freeatbeg)
+    end
+end
+
+BufferWord(v::Union{<:Vector{<:Integer},<:AbstractVector{<:Integer}}) =
+    BufferWord{UInt16}(v)
+BufferWord(sizehint::Integer = 16) =
+    (l = max(sizehint ÷ 2, 8); BufferWord{UInt16}(l, l))
+
+# helper functions for growing storage:
+
+internal_length(bw::BufferWord) = length(bw.storage)
+
+function _growatbeg!(bw::BufferWord, k::Integer)
+    @assert k ≥ 0
+    k == 0 && return bw
+    resize!(bw.storage, internal_length(bw) + k)
+
+    @inbounds for i in 0:length(bw)-1
+        bw.storage[bw.ridx+k-i] = bw.storage[bw.ridx-i]
+    end
+
+    bw.ridx += k
+    bw.lidx += k
+
+    return bw
+end
+
+_growatend!(bw::BufferWord, k::Integer) =
+    resize!(bw.storage, internal_length(bw) + k)
+
+# AbstractWord Interface:
+
+Base.@propagate_inbounds function Base.getindex(bw::BufferWord, n::Integer)
+    @boundscheck checkbounds(bw, n)
+    return @inbounds bw.storage[bw.lidx+n-1]
+end
+
+Base.@propagate_inbounds function Base.setindex!(bw::BufferWord, val, n::Integer)
+    @boundscheck checkbounds(bw, n)
+    return @inbounds bw.storage[bw.lidx+n-1] = val
+end
+Base.length(bw::BufferWord) = length(bw.lidx:bw.ridx)
+
+function Base.push!(bw::BufferWord, k::Integer)
+    if internal_length(bw) == bw.ridx
+        _growatend!(bw, max(length(bw), 16))
+    end
```

Comment:
In case of dynamic arrays, when the current array is full, we double the capacity of such an array. I see that you decided to extend the array by at least 16 cells. Is the reason that the words are usually short?

Reply:
It's just a constant to balance number of additional allocations vs. memory over-conscription. We'd need to tune it or maybe even see if it makes any difference.